continuous-deployment/readme.md

```diff
@@ -1,7 +1,142 @@
+
+* Release Selection: The developer carrying out application testing should have the capability to select which release version to deploy to the testing environment.
+* Environment Readiness - Clean the target cloud environment so it is ready to host a new version of the application.
+* Deployment - Release the application binary(s) to the target cloud environment.
+* Configuration - Applications should be configured in a consistent manner across all your environments. This configuration is applied at the time of deployment. Application Secrets should not be exposed within the runtime environment. Sensitive data like app secrets and certificates should be mastered in Key Vault and sourced internally within the application itself. We encourage 12 Factor principles, especially when it comes to [configuration management](https://12factor.net/config).
+* Data Migration - Pre populate application state and/or data records which is needed for your runtime environment. This may also include test data required for your end-to-end integration test suite.
+* Deployment smoke test. Your smoke test should also verify that your application is pointing to the correct configuration (e.g. production pointing to a UAT Database).
+* Perform any manual or automated acceptance test scenarios.
+* Approve the release gate to promote the application version to the target cloud environment. This promotion should also include the environment's configuration state (e.g. new env settings, feature flags, etc).
+
+#### Pre-production releases
+
+Application release candidates should be deployed to a staging environment similar to production for carrying out final manual/automated tests (_including capacity testing_). Your production and staging / pre-prod cloud environments should be setup at the beginning of your project.
+
+Application warm up should be a quantified measurement that's validated as part of your pre-prod smoke tests.
+
+### Rolling-Back Releases
+
+Your release strategy should account for rollback scenarios in the event of unexpected failures following a deployment.
+
+Rolling back releases can get tricky, especially when database record/object changes occur in result of your deployment (*either inadvertently or intentionally*). If there are no data changes which need to be backed out, then you can simply trigger a new release candidate for the last known production version and promote that release along your CD pipeline.
+
+For rollback scenarios involving data changes, there are several approaches to mitigating this which fall outside of the scope of this guide. Some involve database record versioning, time machining database records / objects, etc. All data files and databases should be backed up prior to each release so they could be restored. The mitigation strategy for this scenario will vary across our projects. The expectation is that this mitigation strategy should be covered as part of your release strategy.
+
+### Zero Downtime Releases
```

Comment:
I think if I don't bring it up, someone else will... What about rings? There was a good article on docs.microsoft.com that explains the basics so think it could be short one liner with link to the article.

Reply:
Added. LMK if the description requires any further refinement.